polynote-runtime/src/main/scala/polynote/runtime/ValueRepr.scala

```diff
@@ -0,0 +1,249 @@
+  */
+final case class MIMERepr(mimeType: String, content: String) extends ValueRepr
+
+/**
+  * A binary representation of a value, encoded in Polynote's format (TODO)
+  */
+final case class DataRepr(dataType: DataType, data: ByteBuffer) extends ValueRepr
+
+/**
+  * An identifier for a "lazy" value. It won't be evaluated until the client asks for it by its handle.
+  * We use Ints for the handle, because JS needs to store the handle, and it can't store Long. We just can't have
+  * more than 2^32^ different handles... I think there would be bigger issues at that point.
+  */
+final case class LazyDataRepr private[polynote](handle: Int, dataType: DataType) extends ValueRepr {
+  def isEvaluated: Option[Boolean] = LazyDataRepr.getHandle(handle).map(_.isEvaluated)
+
+  def release(): Unit = {
+    LazyDataRepr.handles.remove(handle)
+  }
+}
+
+object LazyDataRepr {
+
+  private[polynote] class Handle(val handle: Int, val dataType: DataType, lazyData: => ByteBuffer) {
+    @volatile private var computedFlag: Boolean = false
+
+    lazy val data: ByteBuffer = {
+      computedFlag = true
+      lazyData
+    }
+
+    def isEvaluated: Boolean = computedFlag
+  }
+
+
+  private val handles: ConcurrentHashMap[Int, Handle] = new ConcurrentHashMap()
+  private val nextHandle: AtomicInteger = new AtomicInteger(0)
+  private[polynote] def getHandle(handleId: Int): Option[Handle] = Option(handles.get(handleId))
+
+  def apply(dataType: DataType, value: => ByteBuffer): LazyDataRepr = {
+    val handleId = nextHandle.getAndIncrement()
+    val handle = new Handle(handleId, dataType, value)
+    handles.put(handleId, handle)
+    val repr = LazyDataRepr(handleId, dataType)
+
+    // if the repr object gets GC'ed, we can let the handle go as well
+    Cleaner.create(repr, new Runnable {
+      def run(): Unit = handles.remove(handleId)
+    })
+
+    repr
+  }
+
+}
+
+/**
+  * An identifier for an "updating" value. The server may push one or more binary representations of it over time, and
+  * when there won't be anymore updates, the server will push a message indicating this.
+  */
+final case class UpdatingDataRepr private[polynote](handle: Int, dataType: DataType) extends ValueRepr {
+  def tryUpdate(data: ByteBuffer): Boolean = UpdatingDataRepr.getHandle(handle) match {
+    case Some(h) => h.update(data)
+    case None    => false
+  }
+
+  def update(data: ByteBuffer): Unit =
+    if (!tryUpdate(data))
+      throw new IllegalStateException("Attempt to update a data handle which is already closed")
+
+  def release(): Unit = UpdatingDataRepr.getHandle(handle).foreach(_.release())
+}
+
+object UpdatingDataRepr {
+
+  private[polynote] class Handle(handle: Int, dataType: DataType) {
+    @volatile private var updater: ByteBuffer => Unit = _
+    @volatile private var finalizer: () => Unit = _
+
+    @volatile private var lastState: ByteBuffer = _
+    @volatile private var releaseFlag: Int = 0
```

Comment:
Hmm, I think a comment with some explanation of what the various values of `releaseFlag` mean would be useful here.

Reply:
Yeah, sorry :(